File: IPaImageURLLoader/IPaImageURLLoader/IPaImageURLView.m
//
//  IPaImageURLView.m
//  IPaImageURLLoader
//
//  Created by IPa Chen on 2014/10/12.
//  Copyright (c) 2014年 IPaPa. All rights reserved.
//

#import "IPaImageURLView.h"
#import "IPaImageURLLoader.h"
@implementation IPaImageURLView
{
    id imageObserver;
}
-(void)awakeFromNib
{
    [super awakeFromNib];
    __weak IPaImageURLView *weakSelf = self;
    imageObserver = [[NSNotificationCenter defaultCenter] addObserverForName:IPA_NOTIFICATION_IMAGE_LOADED object:nil queue:[NSOperationQueue mainQueue] usingBlock:^(NSNotification* noti){
        
        if ([noti.userInfo[IPA_NOTIFICATION_KEY_IMAGEID] isEqualToString:weakSelf.imageURL]) {
            [weakSelf setImage:[UIImage imageWithData:[[NSData alloc] initWithContentsOfURL:noti.userInfo[IPA_NOTIFICATION_KEY_IMAGEFILEURL]]]];
        }
    }];
}
-(void)setImageURL:(NSString *)imageURL
{
    [self setImageURL:imageURL withDefaultImage:nil];

    
}
- (void)setImageURL:(NSString *)imageURL withDefaultImage:(UIImage*)defaultImage
{
    imageURL = [imageURL stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding];
    _imageURL = imageURL;
    UIImage *image = [[IPaImageURLLoader defaultLoader] loadImageFromURL:self.imageURL];
    [self setImage:(image == nil)?defaultImage:image];
}
-(void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:imageObserver];
}
@end

File: IPaImageURLLoader/IPaImageURLLoader/IPaImageURLView.swift
//
//  IPaImageURLView.swift
//  IPaImageURLLoader
//
//  Created by IPa Chen on 2015/6/16.
//  Copyright (c) 2015年 AMagicStudio. All rights reserved.
//

import Foundation
import UIKit
@objc open class IPaImageURLView : UIImageView {
    fileprivate var _imageURL:String?
    fileprivate var _highlightedImageURL:String?
    fileprivate var imageObserver:NSObjectProtocol?
    open var imageURL:String? {
        get {
            return _imageURL
        }
        set {
            setImageURL(newValue, defaultImage: nil)
        }
    }
    open var highlightedImageURL:String? {
        get {
            return _highlightedImageURL
        }
        set {
            setHighlightedImageURL(newValue, defaultImage: nil)
        }
    }
    deinit {
        if let imageObserver = imageObserver {
            NotificationCenter.default.removeObserver(imageObserver)
        }
    }
    func createImageObserver () {
        if imageObserver != nil {
            return
        }
        imageObserver = NotificationCenter.default.addObserver(forName: NSNotification.Name(rawValue: IPA_NOTIFICATION_IMAGE_LOADED), object: nil, queue: OperationQueue.main, using: {
            noti in
            
            
            if let userInfo = (noti as NSNotification).userInfo {
                let imageID = userInfo[IPA_NOTIFICATION_KEY_IMAGEID] as! String
                if let imageURL = self.imageURL {
                    if imageID == imageURL {
                        if let data = try? Data(contentsOf: userInfo[IPA_NOTIFICATION_KEY_IMAGEFILEURL] as! URL ) {
                            self.image = UIImage(data: data)
                        }
                    }
                }
                else if let imageURL = self.highlightedImageURL {
                    if imageID == imageURL {
                        if let data = try? Data(contentsOf: userInfo[IPA_NOTIFICATION_KEY_IMAGEFILEURL] as! URL ) {
                            self.highlightedImage = UIImage(data: data)
                        }
                    }
                }
            }
        })
    }
    open func setImageURL(_ imageURL:String?,defaultImage:UIImage?) {
        createImageObserver()
        _imageURL = imageURL
        var image:UIImage?
        if let imageURL = imageURL {
            image = IPaImageURLLoader.sharedInstance.loadImage(url: (imageURL as NSString).addingPercentEncoding(withAllowedCharacters: CharacterSet.urlQueryAllowed)!, imageID: imageURL)
            
        }
        self.image = (image == nil) ? defaultImage :image
    }
    open func setHighlightedImageURL(_ imageURL:String?,defaultImage:UIImage?) {
        createImageObserver()
        _highlightedImageURL = imageURL
        var image:UIImage?
        if let imageURL = imageURL {
            image = IPaImageURLLoader.sharedInstance.loadImage(url: (imageURL as NSString).addingPercentEncoding(withAllowedCharacters: CharacterSet.urlQueryAllowed)!, imageID: imageURL)
            
        }
        self.highlightedImage = (image == nil) ? defaultImage :image
    }
    
}
